server/src/main/java/org/elasticsearch/script/field/WriteField.java

```diff
@@ -0,0 +1,386 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.script.field;
+
+import org.elasticsearch.common.util.Maps;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+public class WriteField implements Field<Object> {
+    protected String path;
+    protected final Supplier<Map<String, Object>> rootSupplier;
+
+    protected Map<String, Object> container;
+    protected String leaf;
```

Comment:
Since we only update container and leaf, does this mean that two write fields operating on the same path may not necessarily reflect the changes of each other?

Reply:
Yes.  If there is a structural change then two write fields may not see the changes with each other.  I think that's fine for an initial implementation.

The ways to solve the problem are to re-resolve on every operation or use a backing data structure for coordination.

Reply:
Do you foresee any issues w/ bwc if we do make the fields reflect changes to other fields? I think long term it's probably most correct to have changes reflected, so I worry if we release this will be a difficult thing to change. I also worry about users just outright considering this a bug, but I do admit for most use cases it'll probably not come up.

Reply:
> Do you foresee any issues w/ bwc if we do make the fields reflect changes to other fields?

Adding functionality is backwards compatible, which is why I want to start with this and then work on adding functionality.

> I also worry about users just outright considering this a bug

Yes, I could see that.  I would say it's a missing feature.  It would be like moving a working directory after a user has `cd`ed into it, it resolved once, at `cd` time and didn't re-resolve.